### DISCUSSION TITLE
Load CMYK Color?

### DISCUSSION THREAD

**Discussion by pha3z** [SixLabors/ImageSharp#2280] [Q&A]
Is it possible to load an image into a buffer with CMYK colors??

I need to write images to CMYK PDF. 

`Image.Load<Cmyk>` causes:

> The type 'SixLabors.ImageSharp.ColorSpaces.Cmyk' cannot be used as type parameter 'TPixel' in the generic type or method 'Image.Load<TPixel>(string)'. There is no boxing conversion from 'SixLabors.ImageSharp.ColorSpaces.Cmyk' to 'SixLabors.ImageSharp.PixelFormats.IPixel<SixLabors.ImageSharp.ColorSpaces.Cmyk>'.

**Comment by JimBobSquarePants** [comment]
No. It's not. All pixel formats are members of the RGB(A) color space. When we load CMYK images we convert them on decode/encode while preserving the embedded color profile. 

All our operations work against `Vector4`. If we used a non-RGB(A) color space that would require converting between color spaces per operation which would be monstrously slow.

**Reply by pha3z** [reply]
> When we load CMYK images we convert them on decode/encode while preserving the embedded color profile.

So you're converting CMKY into RGB and that's what becomes available in the pixel buffer.  I need more explanation on what you mean by preserving the embedded color profile.   Supposing there is an ICC Profile in the image, are you saying you replace the loaded ICC Profile with a new profile that maps RGB to the Profile Connection Space?

I've decided for now I can live without the ability to work with CMYK TIFF directly.  But since I want to read CMYK TIFF and save it out to a new image, I need to be able to attach a proper ICC Profile to make sure the output RGB values map to the same PCS values as the original CMYK values.

Effectively, CMYK with ICC comes in (Loaded with ImageSharp).  I save out RGB (to PDF) with a new ICC  that maintains the correct PCS mapping.

How can I make this happen?

**Comment by pha3z** [comment]
Oh man :(

I totally understand about the operational costs and complexities for the various operations!!

But at least having the ability to use ImageSharp as an Encoder/Decoder to load CMYK images into a CMYK buffer so that a consumer could perform his own operations on them would be super awesome for .NET.  Right now, I don't think there's any other existing library out there for .NET to deal with CMYK... which is sad if you want to do printing/press color-space work. :(

Is there at least support for CMYK TIFF decoding?  I'm wondering what it would take to support decoding into a CMYK buffer and giving that buffer to the consumer for his own operations.

**Reply by dlemstra** [reply]
The Magick.NET library that I created has support for CMYK tiffs so you could use that as an alternative.

**Reply by pha3z** [reply]
I didn't know about Magick.NET!!  It looks like it could work.   I see it requires some native library compilation to support xplat -- I'm targeting Linux, Windows, and macOS (with windows and macOS being most important).   I could make it work if I have to.

I'm still curious to know if decoding of CMYK TIFFs in ImageSharp is supported.  If yes, maybe it wouldn't be a lot of lift to get a buffer with the decoded data and work with it directly.

**Reply by brianpopow** [reply]
> I'm still curious to know if decoding of CMYK TIFFs in ImageSharp is supported. If yes, maybe it wouldn't be a lot of lift to get a buffer with the decoded data and work with it directly.

Yes it is supported since [PR-2268](https://github.com/SixLabors/ImageSharp/pull/2268), but only available in our nightly's so far: [MyGet-Feed](https://www.myget.org/feed/sixlabors/package/nuget/SixLabors.ImageSharp)

**Reply by antonfirsov** [reply]
In theory we could avoid the monstrous operational costs by enabling pixel types to bypass [color conversion in codecs](https://github.com/SixLabors/ImageSharp/blob/8b7eaf257e014b23697649fa46546390ddb6161b/src/ImageSharp/Formats/Jpeg/Components/Decoder/SpectralConverter%7BTPixel%7D.cs#L137). But this would be a far-leading, non-trivial feature.

> decoding into a CMYK buffer and giving that buffer to the consumer for his own operations

If you want to do calculations in Cmyk colorspace, you may also consider processing the pixel buffers of `Image<Rgb24>` , `Image<Rgba32>` or `Image<RgbaVector>` and do the conversions manually after decoding the file into images of these pixel types. The [`ColorSpaceConverter`](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.ColorSpaces.Conversion.ColorSpaceConverter.html) class might be handy for this.
